pytorch_lightning/core/lightning.py

```diff
@@ -1591,8 +1594,12 @@ def to_torchscript(
                 # if no example inputs are provided, try to see if model has example_input_array set
                 if example_inputs is None:
                     example_inputs = self.example_input_array
+                # dicts are not supported, so show the user an error; not raising an error to show the original error
+                if type(example_inputs) == dict:
+                    log.error(f"`example_inputs` should be a Tensor or a tuple of Tensors, but got a dict.")
```

Comment:
```suggestion
                    raise TypeError(f"`example_inputs` should be a Tensor or a tuple of Tensors, but got a dict.")
```

Reply:
@justusschock I on purpose used `log.error`, because the blocker should not be `to_torchscript`, but the logic should stay at `torch.jit.trace`'s side.

Say for example that TorchScript is updated at PyTorch's side (`torch.jit.trace()` does now accept a dict). Now the `to_torchscript()` function can actually support a dict, but Lightning is unnecessarily blocking it. Also, the original error might be more insightful than just a "dict is not accepted".

`log.error` still gives this little bit of extra information to the user, but only acts as a friendly informer, instead of taking over the guard position.
If `torch.jit.trace()` does accept a dict in an update, we only have a nagging error, but not a showstopper.

Reply:
I see. However, currently this is a showstopper and we should treat it like that imo.

if this isn't any longer in the future, we can simply remove this.

Also I'd wonder why we need it at all, if we say, we let `torchscript` handle this...

Reply:
@justusschock This came forth from the discussion with @awaelchli here: https://github.com/PyTorchLightning/pytorch-lightning/pull/4142#discussion_r506950096

> I see. However, currently this is a showstopper and we should treat it like that imo.

If the person reads the docs (generated from `example_inputs: Optional[Union[torch.Tensor, Tuple[torch.Tensor]]]`, it should already be clear that a `dict` should not be used. So I would argue it's not a show stopper, just something that can be encountered depending on how `self.example_input_array` is set.

> Also I'd wonder why we need it at all, if we say, we let `torchscript` handle this...

It could be removed, but since in some cases `self.example_input_array` is set to a `dict`, it's just a nice heads-up to the user if this problem is encountered. Seeing this `log.error` at the beginning of the Traceback is just a bit more friendly.